src/nanoarrow/nanoarrow_device_cuda.c

```diff
@@ -392,6 +417,251 @@ static ArrowErrorCode ArrowDeviceCudaArrayMove(struct ArrowDevice* device_src,
+        device_src, src->children[i], device_dst, dst->children[i], hstream,
+        unknown_buffer_size_count));
+  }
+
+  // ...and for dictionary
+  if (src->dictionary != NULL) {
+    NANOARROW_RETURN_NOT_OK(ArrowDeviceCudaArrayViewCopyBuffers(
+        device_src, src->dictionary, device_dst, dst->dictionary, hstream,
+        unknown_buffer_size_count));
+  }
+
+  return NANOARROW_OK;
+}
+
+static ArrowErrorCode ArrowDeviceCudaCreateStream(struct ArrowDevice* src,
+                                                  struct ArrowDevice* dst, CUstream* out,
+                                                  struct ArrowError* error) {
+  struct ArrowDeviceCudaPrivate* private_data;
+
+  if (dst->device_type == ARROW_DEVICE_CUDA) {
+    private_data = (struct ArrowDeviceCudaPrivate*)dst->private_data;
+  } else if (src->device_type == ARROW_DEVICE_CUDA) {
+    private_data = (struct ArrowDeviceCudaPrivate*)src->private_data;
+  } else {
+    return NANOARROW_OK;
+  }
+
+  NANOARROW_CUDA_RETURN_NOT_OK(cuCtxPushCurrent(private_data->cu_context),
+                               "cuCtxPushCurrent", NULL);
+  CUresult err = cuStreamCreate(out, CU_STREAM_NON_BLOCKING);
+  CUcontext unused;
+  NANOARROW_CUDA_ASSERT_OK(cuCtxPopCurrent(&unused));
+  NANOARROW_CUDA_RETURN_NOT_OK(err, "cuStreamCreate", error);
+
+  return NANOARROW_OK;
+}
+
+static void ArrowDeviceCudaDestroyStream(CUstream hstream) {
+  if (hstream != NANOARROW_CUDA_DEFAULT_STREAM) {
+    NANOARROW_CUDA_ASSERT_OK(cuStreamDestroy(hstream));
+  }
+}
+
+static ArrowErrorCode ArrowDeviceCudaArrayViewCopyInternal(
+    struct ArrowDevice* device_src, struct ArrowDeviceArrayView* src,
+    struct ArrowDevice* device_dst, struct ArrowArray* dst, struct ArrowError* error) {
+  CUresult err;
+  int code;
+  int64_t unknown_buffer_size_count = 0;
+
+  // Create a stream to do the first (and maybe only) round of buffer copying.
+  // The ArrowDeviceArrayView should probably have a sync_event so that we can
+  // force the stream we just created to wait on the event.
+  CUstream hstream = NANOARROW_CUDA_DEFAULT_STREAM;
+  NANOARROW_RETURN_NOT_OK(
+      ArrowDeviceCudaCreateStream(device_src, device_dst, &hstream, error));
+
+  code = ArrowDeviceCudaArrayViewCopyBuffers(device_src, &src->array_view, device_dst,
+                                             dst, hstream, &unknown_buffer_size_count);
+  if (code != NANOARROW_OK) {
+    ArrowDeviceCudaDestroyStream(hstream);
+    ArrowErrorSet(error, "ArrowDeviceCudaArrayViewCopyBuffers() failed");
+    return code;
+  }
+
+  CUstream hstream_data = NANOARROW_CUDA_DEFAULT_STREAM;
+
+  if (unknown_buffer_size_count > 0) {
+    // If buffers with unknown size were encountered, we need to copy all of their
+    // sizes to the CPU, synchronize, then copy all of the buffers once the CPU
+    // knows how many bytes from each buffer need to be copied.
+    code = ArrowDeviceCudaCreateStream(device_src, device_dst, &hstream_data, error);
```

Comment:
+1 here, need to wait on the sync event possibly provided by `src` on `hstream_data`